lib/datadog/kit/events.rb

```diff
@@ -7,41 +7,63 @@ module Datadog
   module Kit
     # Tracking events
     module Events
-      # Attach login event information to the trace
+      APPSEC_LOGIN_SUCCESS_EVENT = 'users.login.success'
+      APPSEC_LOGIN_FAILURE_EVENT = 'users.login.failure'
+
+      # Attach login success event information to the trace
+      #
+      # This method is experimental and may change in the future.
       #
       # @param trace [TraceOperation] Trace to attach data to.
-      # @param success [bool] Whether login operation was successful or not.
       # @param user [Hash<Symbol, String>] User information to pass to
       #   Datadog::Kit::Identity.set_user. Must contain at least :id as key.
       # @param others [Hash<String || Symbol, String>] Additional free-form
       #   event information to attach to the trace.
-      def self.track_login(trace, success:, user:, **others)
-        subtag = success ? 'success' : 'failure'
-        event = "users.login.#{subtag}"
-
-        track(event, trace, **others)
+      def self.track_login_success(trace, user:, **others)
+        track(:appsec, APPSEC_LOGIN_SUCCESS_EVENT, trace, **others)
 
         user_options = user.dup
         user_id = user.delete(:id)
 
         raise ArgumentError, 'missing required key: :user => { :id }' if user_id.nil?
 
-        if success
-          Kit::Identity.set_user(trace, id: user_id, **user_options)
-        else
-          trace.set_tag("appsec.events.users.login.#{subtag}.usr.id", user_id)
-        end
+        Kit::Identity.set_user(trace, id: user_id, **user_options)
+      end
+
+      # Attach login failure event information to the trace
+      #
+      # This method is experimental and may change in the future.
+      #
+      # @param trace [TraceOperation] Trace to attach data to.
+      # @param user_id [String] User id that attempted login
+      # @param user_exists [bool] Whether the user id that did a login attempt exists.
+      # @param others [Hash<String || Symbol, String>] Additional free-form
+      #   event information to attach to the trace.
+      def self.track_login_failure(trace, user_id:, user_exists:, **others)
+        track(:appsec, APPSEC_LOGIN_FAILURE_EVENT, trace, **others)
+
+        raise ArgumentError, 'user_id cannot be nil' if user_id.nil?
+
+        trace.set_tag('appsec.events.users.login.failure.usr.id', user_id)
```

Comment:
Are these simple, unidirectional tags? A.k.a. they get set once, never read, and forwarded along to the Datadog backend?

(As opposed to using tags to facilitate communication between different parts of our instrumentation/products in the Ruby app...)

Reply:
> Are these simple, unidirectional tags?

Unequivocally yes.